File: lib/scanner/plugins/dns/caa.rb
# frozen_string_literal: true

require 'dnsruby'

module Yawast
  module Scanner
    module Plugins
      module DNS
        class CAA < Yawast::Scanner::Base
          include Dnsruby

          def self.caa_info(uri)
            # force DNS resolver to something that works
            # this is done to ensure that ISP resolvers don't get in the way
            # at some point, should probably do something else, but works for now
            @res = Resolver.new({nameserver: ['8.8.8.8']})

            # setup a list of domains already checked, so we can skip them
            @checked = []

            # setup a counter, so we can see if we actually got anything
            @records = 0

            domain = uri.host.to_s

            chase_domain domain

            if @records.zero?
              Yawast::Shared::Output.log_hash 'vulnerabilities',
                                              'missing_caa_records',
                                              {vulnerable: true, record_count: 0}

              puts
              Yawast::Utilities.puts_vuln 'DNS CAA: No records found.'
            else
              Yawast::Shared::Output.log_hash 'vulnerabilities',
                                              'missing_caa_records',
                                              {vulnerable: false, record_count: @records}
            end
          end

          def self.chase_domain(domain)
            while domain != ''
              begin
                # check to see if we've already ran into this one
                return if @checked.include? domain
                @checked.push domain

                # first, see if this is a CNAME. we do this explicitly because
                # some resolvers flatten in an odd way that prevents just checking
                # for the CAA record directly
                cname = get_cname_record(domain)
                if !cname.nil?
                  Yawast::Utilities.puts_info "\t\tCAA (#{domain}): CNAME Found: -> #{cname}"
                  Yawast::Shared::Output.log_value 'dns', 'caa', domain, "CNAME: #{cname}"

                  chase_domain cname.to_s
                else
                  print_caa_record domain
                end
              rescue => e # rubocop:disable Style/RescueStandardError
                Yawast::Utilities.puts_error "\t\tCAA (#{domain}): #{e.message}"
              end

              # strip the leading element off the domain
              domain = domain.partition('.').last
            end
          end

          def self.get_cname_record(domain)
            ans = @res.query(domain, 'CNAME')

            if !ans.answer[0].nil?
              return ans.answer[0].rdata
            else
              return nil
            end
          end

          def self.print_caa_record(domain)
            ans = @res.query(domain, 'CAA')

            if ans.answer.count.positive?
              ans.answer.each do |rec|
                # check for RDATA
                if !rec.rdata.nil?
                  Yawast::Utilities.puts_info "\t\tCAA (#{domain}): #{rec.rdata}"

                  Yawast::Shared::Output.log_append_value 'dns', 'caa', domain, rec.rdata
                  @records += 1
                else
                  Yawast::Utilities.puts_error "\t\tCAA (#{domain}): Invalid Response: #{ans.answer}"
                end
              end
            else
              # no answer, so no records
              Yawast::Utilities.puts_info "\t\tCAA (#{domain}): No Records Found"

              Yawast::Shared::Output.log_value 'dns', 'caa', domain, 'nil'
            end
          end
        end
      end
    end
  end
end

File: yawast/scanner/plugins/dns/caa.py
import copy
from typing import Union, Dict, List, Tuple

from dns import resolver, exception
from dns.resolver import Resolver

from yawast.shared import output

_checked: Dict[str, bool] = {}
_results: List[List[Union[str, List[str]]]] = []


def get_caa(domain: str) -> List[List[Union[str, List[str]]]]:
    global _checked, _results

    # force DNS resolver to something that works
    # this is done to ensure that ISP resolvers don't get in the way
    # at some point, should probably do something else, but works for now
    # (Cloudflare & Google)
    resv: Resolver = resolver.Resolver()
    resv.nameservers = ["1.1.1.1", "8.8.8.8"]

    _chase_domain(domain, resv)

    # reset global data
    ret = copy.deepcopy(_results)
    _results = []
    _checked = {}

    return ret


def _chase_domain(domain: str, resv: Resolver):
    global _checked, _results
    curr: Union[str, None] = domain

    while curr is not None:
        # check to see if we've already ran into this one
        if _checked.get(curr) is not None:
            return

        _checked[curr] = True

        # first, see if this is a CNAME. we do this explicitly because
        # some resolvers flatten in an odd way that prevents just checking
        # for the CAA record directly
        cname = _get_cname(curr, resv)
        if cname is not None:
            _results.append([curr, "CNAME", cname])
            _chase_domain(cname.rstrip("."), resv)
        else:
            caa = _get_caa_records(curr, resv)
            _results.append([curr, "CAA", caa])

        if "." in curr:
            curr = curr.split(".", 1)[-1]
        else:
            curr = None


def _get_caa_records(domain: str, resv: Resolver) -> List[str]:
    records: List[str] = []

    try:
        answers = resv.query(domain, "CAA", lifetime=3)

        for data in answers:
            records.append(data.to_text())
    except (resolver.NoAnswer, resolver.NXDOMAIN, exception.Timeout):
        pass
    except (resolver.NoNameservers, resolver.NotAbsolute, resolver.NoRootSOA):
        output.debug_exception()
        pass

    return records


def _get_cname(domain: str, resv: Resolver) -> Union[str, None]:
    name = None

    try:
        answers = resv.query(domain, "CNAME", lifetime=3)

        for data in answers:
            name = str(data.target)
    except (resolver.NoAnswer, resolver.NXDOMAIN, exception.Timeout):
        pass
    except (resolver.NoNameservers, resolver.NotAbsolute, resolver.NoRootSOA):
        output.debug_exception()
        pass

    return name
